src/compiler/checker.ts

```diff
@@ -4824,6 +4840,45 @@ namespace ts {
             }
         }
 
+        function getDefaultOfTypeParameter(typeParameter: TypeParameter): Type {
+            return hasNonCircularDefault(typeParameter) ? getDefaultFromTypeParameter(typeParameter) : undefined;
+            // const defaultType = getResolvedDefault(typeParameter);
+            // return defaultType !== noConstraintOrDefaultType && defaultType !== circularConstraintOrDefaultType ? defaultType : undefined;
+        }
+
+        function hasNonCircularDefault(type: TypeParameter) {
+            return getResolvedDefault(type) !== circularConstraintOrDefaultType;
+        }
+
+        function getResolvedDefault(typeParameter: TypeParameter) {
+            if (!typeParameter.resolvedDefault) {
+                if (!pushTypeResolution(typeParameter, TypeSystemPropertyName.ResolvedDefault)) {
+                    return circularConstraintOrDefaultType;
+                }
+                const defaultType = getDefaultFromTypeParameter(typeParameter);
+                const type = defaultType && getResolvedDefaultWorker(defaultType);
+                if (!popTypeResolution()) {
+                    return typeParameter.resolvedDefault = circularConstraintOrDefaultType;
+                }
+                typeParameter.resolvedDefault = type || noConstraintOrDefaultType;
+            }
+            return typeParameter.resolvedDefault;
+        }
+
+        function getResolvedDefaultWorker(type: Type): Type {
+            if (type.flags & TypeFlags.TypeParameter) {
+                return getResolvedDefault(<TypeParameter>type);
+            }
+            if (type.flags & TypeFlags.UnionOrIntersection) {
+                const types = (<UnionOrIntersectionType>type).types;
+                const defaultTypes = filter(map(types, getResolvedDefaultWorker), x => x !== circularConstraintOrDefaultType);
+                return type.flags & TypeFlags.Union && defaultTypes.length === types.length ? getUnionType(defaultTypes) :
+                    type.flags & TypeFlags.Intersection && defaultTypes.length ? getIntersectionType(defaultTypes) :

```

Comment:
I reused this from [line 4827](https://github.com/Microsoft/TypeScript/pull/13487/files/442f5408b6ecb4e4b34986bc161ed9037e5e61b5#diff-c3ed224e4daa84352f7f1abcd23e8ccaR4827), above. It does seem like it should also be performing a comparison against types.length as well, however I am unsure as to whether there was a specific reason for 4827 to ignore circularity in an intersection type.